cmake/spidermonkey.cmake

```diff
@@ -21,24 +19,95 @@ endif()
 # This can be set, for example, to the output directly (`release/` or `debug/`)
 # under a local clone of the `spidermonkey-wasi-embedding` repo.
 if (DEFINED ENV{SPIDERMONKEY_BINARIES})
-    set(SM_SOURCE_DIR $ENV{SPIDERMONKEY_BINARIES})
+    set(SM_LIB_DIR $ENV{SPIDERMONKEY_BINARIES})
+    message(STATUS "Using pre-built SpiderMonkey artifacts from local directory ${SM_LIB_DIR}")
 else()
-    CPMAddPackage(NAME spidermonkey-${SM_BUILD_TYPE}
-            URL https://github.com/bytecodealliance/spidermonkey-wasi-embedding/releases/download/rev_${SM_REV}/spidermonkey-wasm-static-lib_${SM_BUILD_TYPE}.tar.gz
-            DOWNLOAD_ONLY YES
+    set(SM_URL https://github.com/bytecodealliance/starlingmonkey/releases/download/libspidermonkey_${SM_TAG}/spidermonkey-static-${SM_BUILD_TYPE}.tar.gz)
+    execute_process(
+            COMMAND curl -s -o /dev/null -w "%{http_code}" ${SM_URL}
+            RESULT_VARIABLE CURL_RESULT
+            OUTPUT_VARIABLE HTTP_STATUS
     )
 
-    set(SM_SOURCE_DIR ${CPM_PACKAGE_spidermonkey-${SM_BUILD_TYPE}_SOURCE_DIR} CACHE STRING "Path to spidermonkey ${SM_BUILD_TYPE} build" FORCE)
+    if (CURL_RESULT EQUAL 0 AND HTTP_STATUS STREQUAL "200")
+        message(STATUS "Using pre-built SpiderMonkey artifacts from ${SM_URL}")
+        CPMAddPackage(NAME spidermonkey-${SM_BUILD_TYPE}
+                URL ${SM_URL}
+                DOWNLOAD_ONLY YES
+        )
+        set(SM_LIB_DIR ${CPM_PACKAGE_spidermonkey-${SM_BUILD_TYPE}_SOURCE_DIR} CACHE STRING "Path to spidermonkey ${SM_BUILD_TYPE} build" FORCE)
+    else()
+        message(STATUS "No pre-built ${SM_BUILD_TYPE} SpiderMonkey artifacts available for tag ${SM_TAG}. Building from source.")
+    endif()
 endif()
 
-set(SM_INCLUDE_DIR ${SM_SOURCE_DIR}/include)
-
 file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/null.cpp "")
-file(GLOB SM_OBJS ${SM_SOURCE_DIR}/lib/*.o)
 
-add_library(spidermonkey STATIC)
-target_sources(spidermonkey PRIVATE ${SM_OBJS} ${CMAKE_CURRENT_BINARY_DIR}/null.cpp)
-target_include_directories(spidermonkey PUBLIC ${SM_INCLUDE_DIR})
-target_link_libraries(spidermonkey PUBLIC ${SM_SOURCE_DIR}/lib/libjs_static.a)
+if (DEFINED SM_LIB_DIR)
+    set(SM_INCLUDE_DIR ${SM_LIB_DIR}/include)
+
+    file(GLOB SM_OBJS ${SM_LIB_DIR}/lib/*.o)
+
+    add_library(spidermonkey STATIC)
+    target_sources(spidermonkey PRIVATE ${SM_OBJS} ${CMAKE_CURRENT_BINARY_DIR}/null.cpp)
+    target_include_directories(spidermonkey PUBLIC ${SM_INCLUDE_DIR})
+    target_link_libraries(spidermonkey PUBLIC ${SM_LIB_DIR}/lib/libjs_static.a)
+else()
+    CPMAddPackage(NAME gecko-source
```

Comment:
Does this support shallow cloning? I think external project has something like `GIT_SHALLOW 1`.

Reply:
Shallow cloning was supposed to be done automatically. Turns out that doesn't actually work in `ExternalProject` itself, so I changed things pretty substantially to make shallow cloning an actual thing.